brave/src/test/java/brave/features/propagation/SecondarySamplingTest.java

```diff
@@ -13,195 +13,344 @@
+ * that the other two systems.
+ *
+ * <h2>Trace forwarder</h2>
+ * Each of the three systems have different capacities, retention rates and potentially different
+ * billing implications. There should be no interference between these systems. For example, if
+ * Zipkin is sampling 1% and while Triage is also on, still only the 1% sampled by B3 should be in
+ * Zipkin. On the other hand, if Triage is off, it should not accidentally get 100% data from Edge.
+ *
+ * <p>The responsibility for this is a zipkin-compatible endpoint, which can route the same data to
+ * one or more systems who want it. We'll call this the trace forwarder. Some examples are
+ * [PitchFork](https://github.com/jeqo/zipkin-forwarder) and [Zipkin
+ * Forwarder](https://github.com/jeqo/zipkin-forwarder). As trace data is completely out-of-band,
+ * any inputs to help decide where to route the data must be in the Zipkin json. For example, if the
+ * forwarder sees a tag relating to triage, it can build a forwarding rule based on this.
+ *
+ * <h2>The application is unaware of the other tracing systems</h2>
+ * It is critical that these tools in no way change the api surface to instrumentation libraries
+ * such as what's used by frameworks like Spring Boot. The fact that there are multiple systems
+ * choosing data differently should not be noticeable by instrumentation. All of these systems use
+ * the same trace and span IDs, which means log correlation is not affected. Sharing instrumentation
+ * and reporting means we are not burdening the application with redundant overhead. It also means
+ * we are not requiring engineering effort to re-instrument each time we add a system.
+ *
+ * <h2>Implementing a sampling overlay in Brave</h2>
+ * Brave assumes there is a primary, propagated sampling decision, captured in process as {@link
+ * TraceContext#sampled()} at the beginning of the trace, and propagated downstream in B3 headers.
+ * This simplifies the question of what to do about Edge who make a constant decision, and Triage,
+ * who make a conditional decision anywhere. Since they are not the primary deciders of the trace,
+ * they cannot affect {@link TraceContext#sampled()}.
+ *
+ * <p>This implies Edge and Triage use secondary, extra fields, to propagate state about their
+ * respective decisions. When Edge or Triage decides to record data, they must set {@link
+ * TraceContext#sampledLocal()} so that recording occurs even if the primary decision is unsampled.
+ * At startup, {@link Tracing.Builder#alwaysReportSpans()} ensures the trace forwarder gets data
+ * regardless of the B3 decision.
+ *
+ * <p>Secondary fields and state management about them is the responsibility of {@link
+ * ExtraFieldPropagation}. Here, you can add an extra field that corresponds to a header sent out of
+ * process. You can also add an {@link ExtraFieldPlugin} to ensure a relevant tag gets to the trace
+ * forwarder.
```

Comment:
This is awesome! Probably worth it to have a copy on Zipkin's wiki

Reply:
good idea I didn't even consider that. this is a massive javadoc :)